openwisp_monitoring/device/api/views.py

```diff
@@ -42,160 +42,17 @@ def has_object_permission(self, request, view, obj):
-                continue
-            for trace in chart['traces']:
-                header.append(self._get_csv_header(chart, trace))
-                columns.append(trace[1])
-        rows = [header]
-        for index, element in enumerate(data.get('x', [])):
-            row = []
-            for column in columns:
-                row.append(column[index])
-            rows.append(row)
-        for chart in histograms:
-            rows.append([])
-            rows.append([chart['title']])
-            # Export value as 0 if it is None
-            for key, value in chart['summary'].items():
-                if chart['summary'][key] is None:
-                    chart['summary'][key] = 0
-            # Sort Histogram on the basis of value in the descending order
-            sorted_charts = sorted(
-                chart['summary'].items(), key=lambda x: x[1], reverse=True
-            )
-            for field, value in sorted_charts:
-                rows.append([field, value])
-        # write CSV to in-memory file object
-        fileobj = StringIO()
-        csv.writer(fileobj).writerows(rows)
-        return fileobj.getvalue()
-
-    def _get_csv_header(self, chart, trace):
-        header = trace[0]
-        return f'{header} - {chart["title"]}'
-
-    def post(self, request, pk):
-        self.instance = self.get_object()
-        self._init_previous_data()
-        self.instance.data = request.data
-        # validate incoming data
-        try:
-            self.instance.validate_data()
-        except ValidationError as e:
-            logger.info(e.message)
-            return Response(e.message, status=status.HTTP_400_BAD_REQUEST)
-        try:
-            # write data
-            self._write(request, self.instance.pk)
-        except ValidationError as e:
-            logger.info(e.message_dict)
-            return Response(e.message_dict, status=status.HTTP_400_BAD_REQUEST)
-        device_metrics_received.send(
-            sender=self.model, instance=self.instance, request=request
-        )
-        return Response(None)
-
-    def _init_previous_data(self):
-        """
-        makes NetJSON interfaces of previous
-        snapshots more easy to access
-        """
-        data = self.instance.data or {}
-        if data:
-            data = deepcopy(data)
-            data['interfaces_dict'] = {}
-        for interface in data.get('interfaces', []):
-            data['interfaces_dict'][interface['name']] = interface
-        self._previous_data = data
-
-    def _write(self, request, pk):
+class MetricChartsMixin:
+    def _write(self, pk, data=None):
         """
         write metrics to database
         """
-        # saves raw device data
-        self.instance.save_data()
-        data = self.instance.data
+        try:
+            # saves raw device data
+            self.instance.save_data()
+            data = self.instance.data
+        except AttributeError:
+            pass
```

Comment:
Can you please explain why this was required? 

The default value for data is `None`. If this part raised an AttributeError, the value of data will still remain `None` and in next lines the code will throw an error while performing `data.get()`.

Reply:
There was a mistake here but I've fixed it in the latest commit